Add BufferedChangeEvent object

jyavenard opened the pull request:
#### 03f910f7ebd4cbeade20a0c7ebee6043979268ec
<pre>
Add BufferedChangeEvent object
<a href="https://bugs.webkit.org/show_bug.cgi?id=256985">https://bugs.webkit.org/show_bug.cgi?id=256985</a>
rdar://109532732

Reviewed by Youenn Fablet.

Pass to `bufferedchange` a dictionary containing the range added and removed
from the last operation on the source buffer.

Fly-by: remove the need for code to be conditional on both ENABLE(MEDIA_SOURCE) and
ENABLE(MANAGED_MEDIA_SOURCE), assume that if the latter is true, then
so is the first, as doing otherwise would be totally nonsensical.

* LayoutTests/media/media-source/media-managedmse-bufferedchange-expected.txt:
* LayoutTests/media/media-source/media-managedmse-bufferedchange.html: Check that object received is of type BufferedChangeEvent
* LayoutTests/media/media-source/mock-managedmse-bufferedchange-expected.txt: Added.
* LayoutTests/media/media-source/mock-managedmse-bufferedchange.html: Added.
* LayoutTests/media/utilities.js:
(timeRangesToString):
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediasource/BufferedChangeEvent.cpp: Copied from Source/WebCore/Modules/mediasource/ManagedSourceBuffer.cpp.
(WebCore::BufferedChangeEvent::BufferedChangeEvent):
(WebCore::BufferedChangeEvent::eventInterface const):
* Source/WebCore/Modules/mediasource/BufferedChangeEvent.h: Copied from Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h.
* Source/WebCore/Modules/mediasource/BufferedChangeEvent.idl: Copied from Source/WebCore/Modules/mediasource/ManagedSourceBuffer.idl.
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.idl:
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.cpp:
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h:
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.idl:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::buffered):
(WebCore::SourceBuffer::sourceBufferPrivateBufferedChanged):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setBufferingPolicy): Change to be conditional on MANAGED_MEDIA_SOURCE
(WebCore::HTMLMediaElement::purgeBufferedDataIfPossible): same as above
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::operator+=): Add alias to `unionWith` method for naming consistency.
(WebCore::PlatformTimeRanges::operator-=): Add operator to process calculation of range removal.
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:

Canonical link: <a href="https://commits.webkit.org/264472@main">https://commits.webkit.org/264472@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6327aeade0442937a7f5d5416377daaa3d0f38ce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7724 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7995 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9365 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7881 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7730 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9985 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7916 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/10755 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7858 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8976 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7093 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9473 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6285 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7032 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14698 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6587 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7438 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7157 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10565 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/7307 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7647 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6233 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/7879 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6975 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6942 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1804 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1845 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11186 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/8091 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7381 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1953 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on LayoutTests/media/utilities.js:
```diff
@@ -78,3 +78,15 @@ function waitForVideoFrameUntil(video, time, cb) {
     return p;
 }
+
+function timeRangesToString(r) {
```
Single letter variables are  terrible

jyavenard replied:
renamed it to `ranges`

eric-carlson commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -1356,11 +1356,21 @@ void SourceBuffer::setShouldGenerateTimestamps(bool flag)
+        removedRanges -= ranges;
+        auto removedTimeRanges = removedRanges.length() ? RefPtr { TimeRanges::create(WTFMove(removedRanges)) } : nullptr;
+
+        queueTaskToDispatchEvent(*this, TaskSource::MediaElement, BufferedChangeEvent::create(WTFMove(addedTimeRanges), WTFMove(removedTimeRanges)));
```
Would it be helpful to assert that at least one of the ranges is not NULL?

jyavenard replied:
yes it would

youennf commented on Source/WebCore/Modules/mediasource/BufferedChangeEvent.cpp:
```diff
@@ -0,0 +1,63 @@
+
+#include "config.h"
+
+#if ENABLE(MANAGED_MEDIA_SOURCE) && ENABLE(MEDIA_SOURCE)
```
It does not make a lot of sense to have ENABLE(MANAGED_MEDIA_SOURCE) without ENABLE(MEDIA_SOURCE).
Can we just write it as `#if ENABLE(MANAGED_MEDIA_SOURCE)`?

jyavenard replied:
The issue is with the GTK build, MANAGED_MEDIA_SOURCE is enabled by default, but they have this thing that can disable MEDIA_SOURCE anyway. So you could end up in a build that has MANAGED_MEDIA_SOURCE but not MEDIA_SOURCE and so it would be broken. But in this particular, I agree, it wouldn't matter.

youennf commented on Source/WebCore/Modules/mediasource/BufferedChangeEvent.idl:
```diff
@@ -0,0 +1,44 @@
+ */
+
+[
+    Conditional=MANAGED_MEDIA_SOURCE&MEDIA_SOURCE,
```
Ditto here and below

youennf commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -1356,11 +1356,23 @@ void SourceBuffer::setShouldGenerateTimestamps(bool flag)
+        ASSERT(addedTimeRanges || removedTimeRanges, "Can't generate an empty dictionary");
+        queueTaskToDispatchEvent(*this, TaskSource::MediaElement, BufferedChangeEvent::create(WTFMove(addedTimeRanges), WTFMove(removedTimeRanges)));
+    }
+    m_buffered = TimeRanges::create(ranges);
```
The usual pattern is to update the value in the task queue that fires the event.
Should we update m_buffered just before firing the buffered change event?

jyavenard replied:
m_buffered needs to be updated in this method as there are part of the spec where the buffered range is updated synchronously (as seen in PR #14175). So for example, you could call MediaSource.endOfStream() and immediately after read the buffered range value. This method is called in both cases, it will actually also update the element's buffered range following the call to `setBufferedDirty(true);` in that method.

Otherwise, updating m_buffered before or after the call to ` queueTaskToDispatchEvent` ends with the same outcome. The queued task won't be run before the end of the current event loop;

youennf commented on Source/WebCore/Modules/mediasource/BufferedChangeEvent.h:
```diff
@@ -0,0 +1,66 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
```
#if ENABLE(MANAGED_MEDIA_SOURCE)?

jyavenard replied:
indeed. thanks

jyavenard committed:
Add BufferedChangeEvent object
https://bugs.webkit.org/show_bug.cgi?id=256985
rdar://109532732

Reviewed by Youenn Fablet.

Pass to `bufferedchange` a dictionary containing the range added and removed
from the last operation on the source buffer.

Fly-by: remove the need for code to be conditional on both ENABLE(MEDIA_SOURCE) and
ENABLE(MANAGED_MEDIA_SOURCE), assume that if the latter is true, then
so is the first, as doing otherwise would be totally nonsensical.

* LayoutTests/media/media-source/media-managedmse-bufferedchange-expected.txt:
* LayoutTests/media/media-source/media-managedmse-bufferedchange.html: Check that object received is of type BufferedChangeEvent
* LayoutTests/media/media-source/mock-managedmse-bufferedchange-expected.txt: Added.
* LayoutTests/media/media-source/mock-managedmse-bufferedchange.html: Added.
* LayoutTests/media/utilities.js:
(timeRangesToString):
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediasource/BufferedChangeEvent.cpp: Copied from Source/WebCore/Modules/mediasource/ManagedSourceBuffer.cpp.
(WebCore::BufferedChangeEvent::BufferedChangeEvent):
(WebCore::BufferedChangeEvent::eventInterface const):
* Source/WebCore/Modules/mediasource/BufferedChangeEvent.h: Copied from Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h.
* Source/WebCore/Modules/mediasource/BufferedChangeEvent.idl: Copied from Source/WebCore/Modules/mediasource/ManagedSourceBuffer.idl.
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.idl:
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.cpp:
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h:
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.idl:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::buffered):
(WebCore::SourceBuffer::sourceBufferPrivateBufferedChanged):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setBufferingPolicy): Change to be conditional on MANAGED_MEDIA_SOURCE
(WebCore::HTMLMediaElement::purgeBufferedDataIfPossible): same as above
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::operator+=): Add alias to `unionWith` method for naming consistency.
(WebCore::PlatformTimeRanges::operator-=): Add operator to process calculation of range removal.
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:

Canonical link: https://commits.webkit.org/264472@main
